### PR TITLE
Last working version

### PR DESCRIPTION
**Pull request by midfies**

### CODE REVIEW

File: src/test_servers.py [hunk 1]
```diff
@@ -1 +1,20 @@
 """Tests for http-server client and server modules."""
+
+
```
Comment (WillWeatherford) [inline src/test_servers.py:3]: Make sure you also test for empty strings (an example of an edge case).

File: src/test_servers.py [hunk 2]
```diff
@@ -1 +1,20 @@
...
+    assert client("Msg") == "Msg"
+
+
+def test_message_same_as_buffer():
+    """Test message is smaller than a buffer."""
+    from client import client
+
+    assert client("MsgEIGHT") == "MsgEIGHT"
+
+
```
Comment (WillWeatherford) [inline src/test_servers.py:16]: You also need a test for a unicode non-ascii character in here too.

File: src/server.py [hunk 3]
```diff
@@ -1 +1,39 @@
...
+    while True:
+        try:
+            conn, addr = server.accept()
+            if conn:
+                return_message = ''
+                buffer_length = 8
+                message_complete = False
+                while not message_complete:
+                    part = conn.recv(buffer_length)
+                    return_message += part.decode('utf8')
```
Comment (WillWeatherford) [inline src/server.py:23]: So you're decoding each chunk as you get it. Since unicode characters can take up multiple bytes, what happens when a multiple-byte unicode character isn't received all the way in one chunk? i.e. the buffer length cuts it in half?

Rather than decoding one chunk at a time, what would be a better approach?

And anyway, for the purpose of this echo server, do you really need to decode the message at the server end, only to re-encode it and send it back?

File: src/server.py [hunk 4]
```diff
@@ -1 +1,39 @@
...
+    address = ('127.0.0.1', 5030)
+    server.bind(address)
+    server.listen(5)
+    print("Waiting for connection...")
+    while True:
+        try:
+            conn, addr = server.accept()
+            if conn:
+                return_message = ''
+                buffer_length = 8
```
Comment (WillWeatherford) [inline src/server.py:19]: buffer length is a great example of a value which should be a global constant.

File: src/client.py [hunk 5]
```diff
@@ -1 +1,36 @@
 """Client.py for python echo server."""
+import socket
+import sys
+
+
+def client(message):
+    """Client side of the http-echo-server."""
+    
+    destination_info = socket.getaddrinfo("127.0.0.1", 5030)
```
Comment (WillWeatherford) [inline src/client.py:9]: Consider some of the values you are using which don't actually change in your program: the IP/port, the buffer length, and the "end of file" string.

These should be made into global constants, since they won't change and are used in multiple places.

File: src/client.py [hunk 6]
```diff
@@ -1 +1,36 @@
...
+    print('CONNECTED!')
+    buffer_length = 8
+    if len(message) % buffer_length == 0:
+        message += 'EOF'
+    client.sendall(message.encode("utf8"))
+    return_message = u''
+    message_complete = False
+    while not message_complete:
+        part = client.recv(buffer_length)
+        return_message += part.decode('utf8')
```
Comment (WillWeatherford) [inline src/client.py:27]: Same as in your server loop, if you try to decode each arbitrarily-sized chunk at a time, you're gonna have a bad time.

Comment (WillWeatherford) [reply src/client.py:27]: Also, it is much more computationally efficient in Python to append chunks of strings into a list, and then use the `str.join` method, rather than to concatenate two chunks at a time.
